p2p/net/conn.go

```diff
@@ -276,21 +276,19 @@ func (c *FormattedConnection) SendSock(m []byte) error {
 		c.wmtx.Unlock()
 		return fmt.Errorf("connection was closed")
 	}
+	c.wmtx.Unlock()
```

Comment:
It might be that the original scope of the lock was to avoid writing to a closed socket. 
I'm not sure it was helpful, because the socket can be closed from the other side or for internal reasons which `c.closed` doesn't cover, so the write operation must handle this condition gracefully anyway.

If so, do we even need to check `c.closed` when entering? 
`SendSock` can be assumed to be a private method (and if so we should lowercase it), and called only when processing `c.messages` until `c.stopSending` returns, in a higher-level function (`SendListener`). 

If there's some benefit that i'm missing to `c.closed` besides protecting against double close and we still need to check it here, the mutex can be upgraded to RW mutex, because there's no reason for concurrent calls to `SendSock` to block each other on this code block.

Reply:
the interface doc https://golang.org/src/io/io.go?s=4279:4319#L94 says
```
// The behavior of Close after the first call is undefined.
// Specific implementations may document their own behavior.
```
so i guess holding a read lock instead will be safer. otoh, how about using a atomic int to hold `c.closed`?

Reply:
Original mutex was to serialize messages, we now have the `sendListener` `messages` queue for that. it was and still used to prevent double closing and to make sure one and only codepath reports the closing of the connection. if everything works fine I don't see anything wrong with the new approach.

Reply:
I renamed `sendSock` to make it a private method. I don't see anything else that needs to change here. I think the mutex should still work fine, as is, to protect against double-close. Let me know if I've missed anything.

Reply:
It still not clear to be why we need to check for `c.closed` when entering `sendSock`.

You've added a secondary, limited-scope protection against double close. So the first check seems to be just to prevent writing to a closed channel, but as I mentioned, it isn't helpful for that anyway. 


 
You've added a limited-scoped check for that in case we get an error and want to close, to prevent double close, 

 (and so to avoid double close), so if the initial check is for avoid writing to a closed channel,  I mentioned

Reply:
It's still unclear to me why do we need to check for `c.closed` when entering `sendSock`.

You've added a secondary, limited-scope protection against the double close. So the first check seems to be just to prevent writing to a closed channel, but as I mentioned, it isn't helpful for that anyway.

Reply:
> otoh, how about using a atomic int to hold c.closed?

If you mean to use it instead of the mutex lock, It would be sufficient for the initial status check (which I don't think we need), but not for preventing double close, because the close operation is not atomic.

Reply:
fixed